### PR TITLE
Fix slow /source API execution

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
I've investigated the problem described in https://github.com/facebook/WebDriverAgent/issues/320 and also similar issues in my application and got the issue fixed by adding this simple verification.

It turns out, that XCTest triggers

```
Enqueue Failure: UI Testing Failure - Failure fetching attributes for element <XCAccessibilityElement: 0x60000025f9e0> Device element: Error Domain=XCTestManagerErrorDomain Code=13 "Error copying attributes -25202" UserInfo={NSLocalizedDescription=Error copying attributes -25202} <unknown> 0 1
```

log record if we try to get visibility attribute for an element snapshot whose width or height are equal to zero. Also, XCTest waits for 15 seconds after this log record appears, which makes _/source_ command execution extremely slow for some applications. I have tested this change on our app and now it takes milliseconds to call this API in comparison to 1 minute without the change.

### CODE REVIEW

**Comment by yaphatak** [comment]
Also lots of people complained and I have seen sometimes that even if element is on screen `isVisible` returns false. @mykola-mokhnach have you seen that? I think we can fix that in same fashion where if element's location is within `rect` of application then its visible rather than depending on properties from `apple` which are not correct.

**Comment by mykola-mokhnach** [comment]
@yaphatak Let's see whether this change fixes the mentioned issue for other users. Anyway, we need an example of such application and logs, where we can prove that visibility attribute provided by Apple does not return correct result.

**Comment by marekcirkos** [comment]
@yaphatak I noticed that proper visibility value is delayed. Tho if you using visibility to to validate presence it should not be a problem. Assuming you have methods like `waitTillVisible`

File: WebDriverAgentLib/Categories/XCUIElement+FBIsVisible.m [hunk 1]
```diff
@@ -28,7 +28,11 @@ @implementation XCElementSnapshot (FBIsVisible)
 
 - (BOOL)fb_isVisible
 {
-  return [(NSNumber *)[self fb_attributeValue:FB_XCAXAIsVisibleAttribute] boolValue];
+  // To avoid known issue "Error copying attributes -25202" in XCTest framework
+  if (self.frame.size.height > 0 && self.frame.size.width > 0) {
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBIsVisible.m]: I know it is aesthetic stuff, but please revert the check please, to make it look as fast return. And more comment here (eg. what you stated in diff description)

**Comment by marekcirkos** [comment]
This looks like magic to me. Feels like we are covering other issue. Super weird, but if it works, it works :)

File: WebDriverAgentLib/Categories/XCUIElement+FBIsVisible.m (hunk 1 above)
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBIsVisible.m]: Nope, something like this

```objc
if (self.frame.size.height < [threshold] && self.frame.size.width < [threshold] ) {
   //Nice comment
    return NO;
}
return [(NSNumber *)[self fb_attributeValue:FB_XCAXAIsVisibleAttribute] boolValue];
```

Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBIsVisible.m]: done

**Comment by marekcirkos** [comment]
Nice catch! Thanks for that!

**Comment by marekcirkos** [comment]
@facebook-github-bot shipit

**Comment by marekcirkos** [comment]
Can you rebase your diff? I got some conflicts when landing.

**Comment by marekcirkos** [comment]
@yaphatak Please rebase this one as well

**Comment by yaphatak** [comment]
@marekcirkos : you meant @mykola-mokhnach 
Also visibility issue that I am seeing is irrespective of time. Even I wait for 30 seconds visibility property is still not true. I guess thats what coming from underlying apple frmaework

**Comment by marekcirkos** [comment]
@facebook-github-bot import

**Comment by bily621** [comment]
I encountered the same problem too.

[debug] [XCUITest] Executing command 'getPageSource'
[JSONWP Proxy] Proxying [POST /source] to [POST http://localhost:8100/session/CB5B75D2-4680-4308-8C84-86D06BE6F810/source] with no body
[debug] [WebDriverAgent] Device: Oct 27 17:18:27 guafeide-iPhone XCTRunner[766] <Warning>: Enqueue Failure: UI Testing Failure - Failure fetching attributes for element <XCAccessibilityElement: 0x12824b220> Device element: Error Domain=XCTestManagerErrorDomain Code=13 "Error copying attributes -25202" UserInfo={NSLocalizedDescription=Error copying attributes -25202} <unknown> 0 1
[debug] [WebDriverAgent] Device: Oct 27 17:18:42 guafeide-iPhone XCTRunner[766] <Warning>: Enqueue Failure: UI Testing Failure - Failure fetching attributes for element <XCAccessibilityElement: 0x12824b990> Device element: Error Domain=XCTestManagerErrorDomain Code=13 "Error copying attributes -25202" UserInfo={NSLocalizedDescription=Error copying attributes -25202} <unknown> 0 1

**Comment by KBhatti** [comment]
@marekcirkos have you been able to run xcode 8.1/Appium 1.6.2 on Travis? I get it to run perfectly fine locally but on Travis-CI wherever I have an xpath I get this error:
[debug] [WebDriverAgent] Sim: Dec 10 01:40:36 Traviss-Mac XCTRunner[7358]: Enqueue Failure: UI Testing Failure - Failure fetching attributes for element <XCAccessibilityElement: 0x600000255120> Device element: Error Domain=XCTestManagerErrorDomain Code=13 "Error copying attributes -25202" UserInfo={NSLocalizedDescription=Error copying attributes -25202} <unknown> 0 1
[debug] [WebDriverAgent] Sim: Dec 10 01:40:51 Traviss-Mac XCTRunner[7358]: Enqueue Failure: UI Testing Failure - Failure fetching attributes for element <XCAccessibilityElement: 0x600000455480> Device element: Error Domain=XCTestManagerErrorDomain Code=13 "Error copying attributes -25202" UserInfo={NSLocalizedDescription=Error copying attributes -25202} <unknown> 0 1
[debug] [WebDriverAgent] Sim: Dec 10 01:41:07 Traviss-Mac XCTRunner[7358]: Enqueue Failure: UI Testing Failure - Failure fetching attributes for element <XCAccessibilityElement: 0x600000455510> Device element: Error Domain=XCTestManagerErrorDomain Code=13 "Error copying attributes -25202" UserInfo={NSLocalizedDescription=Error copying attributes -25202} <unknown> 0 1
[debug] [WebDriverAgent] Sim: Dec 10 01:41:22 Traviss-Mac XCTRunner[7358]: Enqueue Failure: UI Testing Failure - Failure fetching attributes for element <XCAccessibilityElement: 0x60000024a1a0> Device element: Error Domain=XCTestManagerErrorDomain Code=13 "Error copying attributes -25202" UserInfo={NSLocalizedDescription=Error copying attributes -25202} <unknown> 0 1


Also when I run the 'Scripts/bootstraps.sh -d' under webdriveragent.folder on Travis-CI, it just says "Fetching dependencies" and nothing happens. Could this issue be related to the problem i am faching? Not sure what my solution at this point could be. I see that someone suggested to wait for the next release of WDA..when is that?

Thanks

**Comment by marekcirkos** [comment]
@KBhatti WDA does not have any official releases. This is ongoing project. I think  @mykola-mokhnach meant Appium release

**Comment by KBhatti** [comment]
Thanks @marekcirkos I was just wondering if there was some possibility of Appium having a fix for this in the near future soon.

**Comment by montdidier** [comment]
This is tripping me up also. @marekcirkos hopefully Appium can get this incorporated into a release soon.

**Comment by jlipps** [comment]
@montdidier this change has already been released in Appium. Try Appium 1.6.3 for example.

**Comment by KBhatti** [comment]
@jlipps are you sure 1.6.3 has this change?  I have appium 1.6.3/xcode 8.2/iOS 10.2 and I am still getting the error.

**Comment by jlipps** [comment]
@KBhatti then the error is unrelated to this change. We incorporated latest master from WDA like 2 days ago and this change was merged 2 months ago.

**Comment by suhasgatty** [comment]
@jlipps im Using Appium 1.6.4 beta still im getting the same issue as 

[debug] [MJSONWP] Calling AppiumDriver.getPageSource() with args: ["20ef82c0-7e62-4797-98c0-25008e048d0f"]
[debug] [XCUITest] Executing command 'getPageSource'
[debug] [JSONWP Proxy] Proxying [GET /source] to [GET http://localhost:8137/session/D19756E3-1A7C-4C9D-B014-42BB56E476EC/source] with no body

**Comment by mykola-mokhnach** [comment]
@suhasgatty try the visibility workaround described in https://github.com/facebook/WebDriverAgent/issues/372

**Comment by KBhatti** [comment]
@mykola-mokhnach I did tag you on the last comment on #372 
Please respond when you get the chance.

**Comment by wsndy163** [comment]
https://github.com/wsndy163/WebDriverAgent
修复swipe引起的列表卡顿问题

**Pull request closed without merging**